hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImpl.java

```diff
@@ -0,0 +1,385 @@
+import com.hazelcast.client.impl.spi.ClientClusterService;
+import com.hazelcast.client.impl.spi.ClientPartitionService;
+import com.hazelcast.cluster.Address;
+import com.hazelcast.cluster.Cluster;
+import com.hazelcast.cluster.InitialMembershipEvent;
+import com.hazelcast.cluster.InitialMembershipListener;
+import com.hazelcast.cluster.Member;
+import com.hazelcast.cluster.MemberSelector;
+import com.hazelcast.cluster.MembershipEvent;
+import com.hazelcast.cluster.MembershipListener;
+import com.hazelcast.config.ListenerConfig;
+import com.hazelcast.internal.cluster.MemberInfo;
+import com.hazelcast.internal.cluster.impl.MemberSelectingCollection;
+import com.hazelcast.internal.nio.ClassLoaderUtil;
+import com.hazelcast.internal.nio.Connection;
+import com.hazelcast.internal.util.Clock;
+import com.hazelcast.internal.util.ExceptionUtil;
+import com.hazelcast.internal.util.UuidUtil;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.partition.PartitionLostListener;
+import com.hazelcast.spi.exception.TargetDisconnectedException;
+
+import javax.annotation.Nonnull;
+import java.net.InetSocketAddress;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.EventListener;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.hazelcast.internal.util.Preconditions.checkNotNull;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.unmodifiableSet;
+
+/**
+ * The {@link ClientClusterService}  and {@link ClientPartitionService} implementation.
+ */
+public class ClientClusterServiceImpl implements ClientClusterService {
+
+    private static final int INITIAL_MEMBERS_TIMEOUT_SECONDS = 120;
+
+    private static final MemberListSnapshot EMPTY_SNAPSHOT =
+            new MemberListSnapshot(-1, new LinkedHashMap<>(), Collections.emptySet());
+    private final HazelcastClientInstanceImpl client;
+
+    private final AtomicReference<MemberListSnapshot> memberListSnapshot = new AtomicReference<>(EMPTY_SNAPSHOT);
+    private final ConcurrentMap<UUID, MembershipListener> listeners = new ConcurrentHashMap<>();
+    private final Object clusterViewLock = new Object();
+    private final Set<String> labels;
+    private final ILogger logger;
+    private final ClientConnectionManager connectionManager;
+    private volatile CountDownLatch initialListFetchedLatch = new CountDownLatch(1);
+
+    private static final class MemberListSnapshot {
+        private final int version;
+        private final LinkedHashMap<Address, Member> members;
+        private final Set<Member> memberSet;
```

Comment:
This was an optimization to avoid recreating memberset for each event. Given that we have seperated partition view event and member list event, an event potentially will mean that memberlist has changed and we will recreate the list anyway. So this optimization is not needed any more.  I will change that